### ISSUE TITLE
Astro: Queue full Error at Midnight

### ISSUE THREAD

**Issue by spy0r** [eclipse/smarthome#3809]
Hi,

i discovered the following errors this night:
```
2017-07-06 00:00:00.054 [ERROR] [thome.binding.astro.internal.job.Job] - Queue full
java.lang.IllegalStateException: Queue full
	at java.util.AbstractQueue.add(AbstractQueue.java:98)[:1.8.0_131]
	at org.eclipse.smarthome.binding.astro.handler.AstroThingHandler.addJobToQueue(AstroThingHandler.java:314)[188:org.eclipse.smarthome.binding.astro:0.9.0.b5]
	at org.eclipse.smarthome.binding.astro.internal.job.Job.schedule(Job.java:58)[188:org.eclipse.smarthome.binding.astro:0.9.0.b5]
	at org.eclipse.smarthome.binding.astro.internal.job.Job.scheduleSunPhase(Job.java:156)[188:org.eclipse.smarthome.binding.astro:0.9.0.b5]
	at org.eclipse.smarthome.binding.astro.internal.job.DailyJobSun.run(DailyJobSun.java:89)[188:org.eclipse.smarthome.binding.astro:0.9.0.b5]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_131]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_131]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)[:1.8.0_131]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)[:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748)[:1.8.0_131]
2017-07-06 00:00:00.056 [ERROR] [thome.binding.astro.internal.job.Job] - Queue full
java.lang.IllegalStateException: Queue full
	at java.util.AbstractQueue.add(AbstractQueue.java:98)[:1.8.0_131]
	at org.eclipse.smarthome.binding.astro.handler.AstroThingHandler.addJobToQueue(AstroThingHandler.java:314)[188:org.eclipse.smarthome.binding.astro:0.9.0.b5]
	at org.eclipse.smarthome.binding.astro.internal.job.Job.schedule(Job.java:58)[188:org.eclipse.smarthome.binding.astro:0.9.0.b5]
	at org.eclipse.smarthome.binding.astro.internal.job.Job.scheduleSunPhase(Job.java:156)[188:org.eclipse.smarthome.binding.astro:0.9.0.b5]
	at org.eclipse.smarthome.binding.astro.internal.job.DailyJobSun.run(DailyJobSun.java:90)[188:org.eclipse.smarthome.binding.astro:0.9.0.b5]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)[:1.8.0_131]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0_131]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)[:1.8.0_131]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)[:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0_131]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0_131]
	at java.lang.Thread.run(Thread.java:748)[:1.8.0_131]
```

Openhab is the stable 2.1 release with the resulting astro 0.9.0.b5 binding.

Please see also this thread: https://community.openhab.org/t/astro-binding-queue-full/31123

Thanks

**Comment by mherwege** [comment]
I now have the same error since last night. My 2.1 release has been running for a number of says without touching it. It is the first time the error appears in the logs.

**Comment by sjka** [comment]
Thanks, I can also reproduce it easily (by running the midnight job frequently I didn't even have to wait too long 😄 )

**Comment by sjka** [comment]
Well, the problem is... The astro binding "remembers" in a queue which jobs (like e.g. the events etc) it had scheduled, just in case it gets stopped so that it can clean them up. This queue has a [hard limit of 200](https://github.com/eclipse/smarthome/blob/ref-0.9.0.b5/extensions/binding/org.eclipse.smarthome.binding.astro/src/main/java/org/eclipse/smarthome/binding/astro/handler/AstroThingHandler.java#L63) entries. However, it does not care to remove a job from the queue once the job has been executed. Therefore the queue just fills up until it explodes. What a pity 😢 

The implementation had been changed already, but there is not yet an OH snapshot release which includes it. So upgrading won't help. Restarting your OH will buy you some time though until an updated astro binding is available.

**Comment by spy0r** [comment]
But as i can see this has no impact on any functionality right now - So it's "just" what you mentioned resulting in a log entry, but no "real" problem in the functionality, right?

in which cases does the full queue cause problems when it can not be filled up with the right data?

**Comment by sjka** [comment]
To me it looks like in case of such an error the new daily jobs won't be scheduled (see [Job.java](https://github.com/eclipse/smarthome/blob/ref-0.9.0.b5/extensions/binding/org.eclipse.smarthome.binding.astro/src/main/java/org/eclipse/smarthome/binding/astro/internal/job/Job.java#L58)), so e.g. if you have rules triggered by astro related events they will not be run anymore. 
However, the repeatedly running jobs like the "position" ones keep running, so the sun/moon positions keep updating themselves.

**Comment by mhilbush** [comment]
I can confirm that the astro events are not running, as I have several rules set to trigger on those events, and those rules are not running.  

Shouldn't a start/stop of the astro binding also resolve the issue, rather than restarting OH?

**Comment by spy0r** [comment]
Okay, then i experience "nothing than the logs" because i don't trigger any rule with the items

**Comment by sjka** [comment]
> Shouldn't a stop/start of the astro binding also resolve the issue, rather than restarting OH?

Right, that would also do the trick. Good point, thanks!

**Comment by sjka** [comment]
Closing it as it was fixed by #3776
